File: NodeJs_TypeScript_JavaScript_Server/Project_NodeJs_VSCode/src/base/common/net/tcp/connection/TcpCryptoConnection.ts
import { TcpRequestBlocksConnection } from './TcpRequestBlocksConnection';

import * as net from "net";

import { StreamBuffer } from "base/buffer/StreamBuffer";
import { NetMsgsBlock } from 'base/common/net/NetMsgsBlock';
import { IEncryptCipher } from 'base/crypto/cipher/IEncryptCipher';
import { IDecryptCipher } from 'base/crypto/cipher/IDecryptCipher';

export class TcpCryptoConnection extends TcpRequestBlocksConnection {

  public encryptionCipher: IEncryptCipher = null;
  public decryptionCipher: IDecryptCipher = null;

  private isEnabledAutoCryptoRead: boolean = false;
  private isEnabledAutoCryptoWrite: boolean = false;
  
  public constructor(socket: net.Socket) {
    super(socket);
    
    //void
  }

  public enabledAutoCryptoRead(): void {
    this.isEnabledAutoCryptoRead = true;
}

public enabledAutoCryptoWrite(): void {
    this.isEnabledAutoCryptoWrite = true;
}

  protected extractConfirmedReceivedBlock(blockLen: number): StreamBuffer {
    let blockNetBuffer: StreamBuffer = super.extractConfirmedReceivedBlock(blockLen);
    
    if(this.isEnabledAutoCryptoRead) {
        // Decrypt.
        // The first 2 int values are not encrypted.
        
        let blockBuffer: Buffer = blockNetBuffer.getBuffer();
        
        let cryptoOffset: number = 4 + 4;
        let cryptoDataLen: number = blockBuffer.length  - cryptoOffset;
        
        this.decryptionCipher.processBytes(
          blockBuffer.buffer, cryptoOffset, 
          blockBuffer.buffer, cryptoOffset, 
          cryptoDataLen
        );
      }
      
      return blockNetBuffer;
  }

  public sendBlock(blockDataNetBuffer: NetMsgsBlock): void {
    if(this.isEnabledAutoCryptoWrite) {
      // Encrypt.
      // The first 2 int values are not encrypted.
      
      let blockBuffer: Buffer = blockDataNetBuffer.getBuffer();
      
      let cryptoOffset: number = 4 + 4;
      let cryptoDataLen: number = blockBuffer.length  - cryptoOffset;

      this.encryptionCipher.processBytes(
          blockBuffer.buffer, cryptoOffset, 
          blockBuffer.buffer, cryptoOffset, 
          cryptoDataLen
      );
    }
    
    super.sendBlock(blockDataNetBuffer);
  }
  
}

File: CPP_Joc_Windows_Android/SH_Client_Win_Cpp_Cmake/App/src/base/net/tcp/TcpCryptoConnection.cpp
#include "TcpCryptoConnection.h"
#include <assert.h>
#include <base/MM.h>

using namespace base::Net;

TcpCryptoConnection::TcpCryptoConnection(std::shared_ptr<Socket>& socket)
    : super(socket)
{
    //void
}

void TcpCryptoConnection::enabledAutoCryptoRead() {
    isEnabledAutoCryptoRead = true;
}

void TcpCryptoConnection::enabledAutoCryptoWrite() {
    isEnabledAutoCryptoWrite = true;
}

std::shared_ptr<base::StreamBuffer> TcpCryptoConnection::extractConfirmedReceivedBlock(int blockLen) {
    std::shared_ptr<base::StreamBuffer> blockNetBuffer = super::extractConfirmedReceivedBlock(blockLen);

    if(isEnabledAutoCryptoRead) {
        // Decrypt.
        // The first 2 int values are not encrypted.

        assert(false);

//        asdA3;
//        char* blockBuffer = blockNetBuffer->getBuffer();
//
//        int cryptoOffset = 4 + 4;
//        int cryptoDataLen = blockBuffer.capacity() - cryptoOffset;
//
//        // Create a temporary buffer.
//        byte[] tmpBuff = new byte[cryptoDataLen];
//        JavaUtil.byteArrayCopy(blockBuffer.array(), cryptoOffset, tmpBuff, 0, cryptoDataLen);
//
//        CryptoUtil.decryptBuffer(tmpBuff, decipher);
//
//        // copy the temporary buffer back to the original one.
//        JavaUtil.byteArrayCopy(tmpBuff, 0, blockBuffer.array(), cryptoOffset, cryptoDataLen);
    }

    return blockNetBuffer;
}

void TcpCryptoConnection::sendBlock(std::shared_ptr<NetMsgsBlock> blockDataNetBuffer) {
    if(isEnabledAutoCryptoWrite) {
        // Encrypt.
        // The first 2 int values are not encrypted.

        assert(false);

//        asdA3;
//        ByteBuffer blockBuffer = blockDataNetBuffer.getBuffer();
//
//        int cryptoOffset = 4 + 4;
//        int cryptoDataLen = blockBuffer.capacity()  - cryptoOffset;
//
//        // Create a temporary buffer.
//        byte[] tmpBuff = new byte[cryptoDataLen];
//        JavaUtil.byteArrayCopy(blockBuffer.array(), cryptoOffset, tmpBuff, 0, cryptoDataLen);
//
//        CryptoUtil.encryptBuffer(tmpBuff, cipher);
//
//        // copy the temporary buffer back to the original one.
//        JavaUtil.byteArrayCopy(tmpBuff, 0, blockBuffer.array(), cryptoOffset, cryptoDataLen);
    }

    super::sendBlock(blockDataNetBuffer);
}

std::shared_ptr<base::crypto::IEncryptCipher> TcpCryptoConnection::getEncryptionCipher() {
    return encryptionCipher;
}

void TcpCryptoConnection::setEncryptionCipher(std::shared_ptr<base::crypto::IEncryptCipher> encryptionCipher) {
    this->encryptionCipher = encryptionCipher;
}

std::shared_ptr<base::crypto::IDecryptCipher> TcpCryptoConnection::getDecryptionCipher() {
    return decryptionCipher;
}

void TcpCryptoConnection::setDecryptionCipher(std::shared_ptr<base::crypto::IDecryptCipher> decryptionCipher) {
    this->decryptionCipher = decryptionCipher;
}

TcpCryptoConnection::~TcpCryptoConnection() {
    //void
}
